### PR TITLE
fuse travis-ci

### PR DESCRIPTION
**Pull request by thelinuxkid**
Tests passed https://travis-ci.org/thelinuxkid/go-ipfs/builds/98464151.

### CODE REVIEW

File: core/core.go [hunk 1]
```diff
@@ -471,9 +471,10 @@ func (n *IpfsNode) loadBootstrapPeers() ([]peer.PeerInfo, error) {
 // uses it to instantiate a routing system in offline mode.
 // This is primarily used for offline ipns modifications.
 func (n *IpfsNode) SetupOfflineRouting() error {
-	err := n.LoadPrivateKey()
-	if err != nil {
-		return err
+	if n.PrivateKey == nil {
+		if err := n.LoadPrivateKey(); err != nil {
```
Comment (whyrusleeping) [inline core/core.go]: maybe instead of having this, we should probably just make `LoadPrivateKey` not fail if called while the key is already loaded

**Comment by thelinuxkid** [comment]
Yes. I was afraid to break something but that makes sense. Updated and tests are passing locally.

**Comment by thelinuxkid** [comment]
Working on a rebase for this

**Comment by jbenet** [comment]
Great thanks!

**Comment by thelinuxkid** [comment]
Rebased. I am getting unrelated errors on TestFilePersistence (https://travis-ci.org/thelinuxkid/go-ipfs/jobs/102521435#L252) and TestMultipleDirs (https://travis-ci.org/thelinuxkid/go-ipfs/jobs/102521435#L255).  I will try to take a look at them this weekend.

**Comment by rht** [comment]
@thelinuxkid the travis-ci err has been logged in #2002. I think this just need a rerun.

File: core/core.go [hunk 2]
```diff
@@ -434,12 +434,12 @@ func (n *IpfsNode) loadID() error {
 }
 
 func (n *IpfsNode) LoadPrivateKey() error {
-	if n.Identity == "" || n.Peerstore == nil {
-		return errors.New("loaded private key out of order.")
+	if n.PrivateKey != nil {
+		return nil
```
Comment (jbenet) [inline core/core.go]: im curious why this had to be added? is there a race condition somewhere?

File: .travis/before_install-linux.sh [hunk 3]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+set -ev
+
+sudo apt-get install -qq pkg-config fuse
+sudo modprobe fuse
+sudo chmod 666 /dev/fuse
+sudo chown root:$USER /etc/fuse.conf
```
Comment (jbenet) [inline .travis/before_install-linux.sh]: Will this make CI take much longer? or can travis reuse an image instead of installing FUSE every single time?

Comment (thelinuxkid) [inline .travis/before_install-linux.sh]: I haven't seen a noticeable change. The tests take much much longer than the setup.

File: core/core.go (hunk 2 above)
Comment (thelinuxkid) [inline core/core.go]: See @whyrusleeping's response to my first comment in this PR.

File: .travis/before_install-linux.sh (hunk 3 above)
Comment (jbenet) [inline .travis/before_install-linux.sh]: test times jumped up by 10min in linux and 3min in osx. (they all used to run ~4min (when they dont get stalled))

File: core/core.go (hunk 2 above)
Comment (jbenet) [inline core/core.go]: ah makes sense

**Comment by jbenet** [comment]
- I re-ran tests, the travis go one keeps failing... odd.
- Ooof, the times bumped up to 14min in linux and 7min in osx. rough!
- It is very useful to have fuse though. 
- It's silly for travis-ci to force install every time, i think we should contact them about using dockerfiles so that the FUSE image can be cached forever.

@whyrusleeping and @chriscool could you take a look at these run outputs and confirm all looks good? (i.e. testing fuse as we expect it to be tested, etc):
- https://travis-ci.org/ipfs/go-ipfs/jobs/102976911
- https://travis-ci.org/ipfs/go-ipfs/jobs/102976913

**Comment by chriscool** [comment]
My concern with this is that docker is not available so t0300 is not run:

```
*** t0300-docker-image.sh ***

./t0300-docker-image.sh

lib/test-lib.sh: line 45: type: docker: not found

# TEST_VERBOSE=1

# TEST_NO_FUSE=

# TEST_EXPENSIVE=1

[36m1..0 # SKIP skipping docker tests, docker not available(B[m
```

Maybe we could have the tests run in both configurations so that both fuse tests and tests with docker can be run?

**Comment by jbenet** [comment]
> Maybe we could have the tests run in both configurations so that both fuse tests and tests with docker can be run?

yeah that sounds good. do we know how to set this up?

i wish we could pay travis money to add to the number of workers we get. sigh. they keep not letting me pay them!!

File: .travis/before_install-linux.sh (hunk 3 above)
Comment (Kubuxu) [inline .travis/before_install-linux.sh]: Installation shouldn't take long as Travis uses local mirrors.

**Comment by chriscool** [comment]
@jbenet I am not sure it is possible to have both kind of tests.

I am testing things on this PR to see if it's possible: https://github.com/ipfs/go-ipfs/pull/2241

**Comment by jbenet** [comment]
@chriscool we should be able to have two different builders with different env flags. the ipfsbot account should let you experiment with travis, it has perms there.

**Comment by chriscool** [comment]
@jbenet according to the result of PR #2241, it looks like when "dist: trusty" is set in the .travis.yml file, then docker is not available. So it looks like we might just want to remove "dist: trusty" from .travis.yml, and then we don't need two different builders. One sharness build will test both fuse and docker.

@thelinuxkid was there a reason you added, or uncommented, "dist: trusty" into .travis.yml?

**Comment by thelinuxkid** [comment]
@jbenet fuse was added on trusty build: https://github.com/ipfs/go-ipfs/issues/2023, https://github.com/travis-ci/travis-ci/issues/1100#issuecomment-160169121

**Comment by thelinuxkid** [comment]
FWIT fuse module is already installed on trusty: https://github.com/thelinuxkid/go-ipfs/commit/764e756a4820784232351206ef377afc0b348dff, https://travis-ci.org/thelinuxkid/go-ipfs/builds/104456298.

**Comment by chriscool** [comment]
@thelinuxkid yeah but it looks like if we don't specify "dist: trusty", we can have both fuse and also docker.

**Comment by thelinuxkid** [comment]
I am pro containers. Specially if it eliminates the needs to depend on travis' build images. But, I know nothing of its support in travis. Having said that, I removed the explicit trusty distribution (https://travis-ci.org/thelinuxkid/go-ipfs/builds/104545197) and travis still used trusty (https://travis-ci.org/thelinuxkid/go-ipfs/jobs/104545198#L14). Is that some sort of confusion...bug...is trusty now the default?

**Comment by thelinuxkid** [comment]
BTW except for https://github.com/ipfs/go-ipfs/issues/2214 (which is resolved but needs to be implemented) and stalls, I am not getting any more travis errors.

**Comment by chriscool** [comment]
@thelinuxkid yeah, it looks like Docker works now on Travis. So now this PR LGTM! Thanks!

**Comment by chriscool** [comment]
@whyrusleeping or @jbenet could you take a look at merging this?
The test failures are not related and have been seen elsewhere.

**Comment by rht** [comment]
This needs a rebase from @thelinuxkid

**Comment by jbenet** [comment]
Or Anyone can rebase and PR again

**Comment by thelinuxkid** [comment]
Rebased. Travis tests failed in a weird way. Rerunning. Will report.

**Comment by chriscool** [comment]
There is this go test failure:

```
--- FAIL: TestFilePersistence (0.22s)
	ipns_test.go:189: Closed, opening new fs
	ipns_test.go:195: open /tmp/fusetest063103241/local/atestfile: no such file or directory
--- FAIL: TestMultipleDirs (0.24s)
	ipns_test.go:206: make a top level dir
	ipns_test.go:212: write a file in it
	ipns_test.go:217: sub directory
	ipns_test.go:222: file in that subdirectory
	ipns_test.go:228: closing mount, then restarting
	ipns_test.go:88: stat /tmp/fusetest677160995/local/test1: no such file or directory
FAIL
FAIL	github.com/ipfs/go-ipfs/fuse/ipns	7.021s
```

but it's known and not related.

**Comment by thelinuxkid** [comment]
@chriscool yep. It's addressed here: https://github.com/ipfs/go-ipfs/issues/2214. Everything else looks good.

**Comment by whyrusleeping** [comment]
this looks good to me, mind rebasing on latest master?

**Comment by chriscool** [comment]
@thelinuxkid do you want me to rebase on master?

**Comment by thelinuxkid** [comment]
@chriscool just rebased. Thanks.

**Comment by thelinuxkid** [comment]
Still getting the test failures from #2214. I will try to work on that tomorrow and/or Friday if no one else can pick it up.

**Comment by rht** [comment]
Why did the #2214 happen twice specifically here?

**Comment by thelinuxkid** [comment]
@rht Because this enables the fuse tests

**Comment by thelinuxkid** [comment]
I rebased and the travis tests were failing. So I `reset --hard upstream/master` and pushed to see if I had broken anything. That closed this PR because Github detected no new commits. I pushed again with the changes this time but it won't reopen. Sorry. I should have tested locally instead of relaying on travis.

**Pull request closed without merging**